modules/core/src/main/java/org/apache/ignite/internal/jdbc2/JdbcBinaryBuffer.java

```diff
@@ -0,0 +1,463 @@
+        return new JdbcBinaryBuffer(arr, 0, arr.length, false);
+    }
+
+    /**
+     * Create empty buffer which starts working in the read-write mode.
+     */
+    public static JdbcBinaryBuffer createReadWrite() {
+        return new JdbcBinaryBuffer(new byte[MIN_CAP], 0, 0, false);
+    }
+
+    /**
+     * Private constructor.
+     *
+     * @param arr The byte array to be wrapped.
+     * @param off The offset to the first byte to be wrapped.
+     * @param length The length in bytes of the data to be wrapped.
+     * @param isReadOnly The read-only flag.
+     */
+    private JdbcBinaryBuffer(byte[] arr, int off, int length, boolean isReadOnly) {
+        this.arr = arr;
+        this.off = off;
+        this.length = length;
+        this.isReadOnly = isReadOnly;
+    }
+
+    /**
+     * Create shallow read-only copy of this buffer.
+     */
+    public JdbcBinaryBuffer shallowCopy() {
+        return new JdbcBinaryBuffer(arr, off, length, true);
+    }
+
+    /**
+     * Provide InputStream through which the data can be read starting from the
+     * begining.
+     *
+     * <p>Stream is not limited meaning that it would return any new data
+     * written to the buffer after stream creation.
+     *
+     * @return InputStream instance.
+     */
+    public InputStream getInputStream() {
+        return new BufferInputStream();
+    }
+
+    /**
+     * Get copy of the buffer data as byte array.
+     *
+     * @return Byte array containing buffer data.
+     */
+    public byte[] getBytes() {
+        byte[] bytes = new byte[length];
+
+        read(0, bytes, 0, length);
+
+        return bytes;
+    }
+
+    /**
+     * Provide OutputStream through which the data can be written to buffer starting from
+     * the (zero-based) {@code pos} position.
+     *
+     * @param pos The zero-based offset to the first byte to be written. Must not be negative
+     *            or greater than total count of bytes in buffer.
+     *
+     * @return OutputStream instance.
+     */
+    OutputStream getOutputStream(int pos) {
+        return new BufferOutputStream(pos);
+    }
+
+    /**
+     * Provide InputStream through which the no more than {@code len} bytes can be read
+     * from buffer starting from the specified zero-based position {@code pos}.
+     *
+     * @param pos The zero-based offset to the first byte to be retrieved. Must not be negative
+     *            or greater than total count of bytes in buffer.
+     * @param len The length in bytes of the data to be retrieved. Must not be negative.
+     * @return InputStream instance.
+     */
+    InputStream getInputStream(int pos, int len) {
+        return new BufferInputStream(pos, len);
+    }
+
+    /**
+     * Truncate data in this buffer to specified length.
+     *
+     * @param len New length.
+     */
+    void truncate(int len) {
+        byte[] newArr = new byte[Math.max(MIN_CAP, len)];
+
+        U.arrayCopy(arr, off, newArr, 0, len);
+
+        arr = newArr;
```

Comment:
Let's create a single method that copy data into new array and set `isReadOnly = false` to make sure we do it consistent.

All methods that writes to the `arr` must check `isReadOnly` flag, also.